File: src/331_Verify_Preorder_Serialization_of_a_Binary_Tree.py
 # 331_Verify_Preorder_Serialization_of_a_Binary_Tree
 # https://leetcode.cn/problems/verify-preorder-serialization-of-a-binary-tree/
 
# 逆向栈
class Solution(object):
    def isValidSerialization(self, preorder):
        stack = []
        
        for node in preorder.split(','):
            stack.append(node)
            
            while len(stack) >= 3 and stack[-1] == stack[-2] == '#' and stack[-3] != '#':
                stack.pop(), stack.pop(), stack.pop()
                stack.append('#')
        return len(stack) == 1 and stack.pop() == '#'


# 槽位思想
class Solution(object):
    def isValidSerialization(self, preorder):
        nodes = preorder.split(',')
        diff = 1
        
        for node in nodes:
            diff -= 1
            
            if diff < 0:
                return False
            
            if node != '#':
                diff += 2
                
        return diff == 0

File: src/331_Verify_Preorder_Serialization_of_a_Binary_Tree.rs
// # 331_Verify_Preorder_Serialization_of_a_Binary_Tree
// # https://leetcode.cn/problems/verify-preorder-serialization-of-a-binary-tree/

impl Solution {
    fn is_valid_serialization(preorder: String) -> bool {
        let mut slots = 1;
        let nodes = preorder.split(',');

        for node in nodes {
            slots -= 1;
            if slots < 0 {
                return false;
            }
            if node != "#" {
                slots += 2;
            }
        }

        slots == 0
    }
}


impl Solution {
    pub fn is_valid_serialization(preorder: String) -> bool {
        let mut remain = 1;
        for node in preorder.split(',') {
            if remain == 0 {
                return false;
            }
            if node.parse::<i32>().is_err() {
                remain -= 1;
            } else {
                remain += 1;
            }
        }
        remain == 0
    }
}

